File: python/word_break.py
class Solution:
    def wordBreak(self, s: str, wordDict: list[str]) -> bool:
        dp: list[bool] = [False] * (len(s) + 1)
        dp[len(s)] = True
        for i in range(len(s) - 1, -1, -1):
            for w in wordDict:
                if i + len(w) <= len(s) and s[i : i + len(w)] == w:
                    dp[i] = dp[i + len(w)]
                if dp[i]:
                    break

        return dp[0]


def test_wordBreak_case_1():
    # arrange
    s: str = "leetcode"
    wordDict: list[str] = ["leet", "code"]
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.wordBreak(s, wordDict)

    # assert
    assert actual == expected


def test_wordBreak_case_2():
    # arrange
    s: str = "applepenapple"
    wordDict: list[str] = ["apple", "pen"]
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.wordBreak(s, wordDict)

    # assert
    assert actual == expected


def test_wordBreak_case_3():
    # arrange
    s: str = "catsandog"
    wordDict: list[str] = ["cats", "dog", "sand", "and", "cat"]
    expected: bool = False

    # act
    solution = Solution()
    actual = solution.wordBreak(s, wordDict)

    # assert
    assert actual == expected

File: rust/src/code/word_break.rs
struct Solution {}

impl Solution {
    // NOTE: time complexity O(n * m), space complexity O(n), where n is the length of s and m is
    // the length of word_dict
    pub fn word_break(s: String, word_dict: Vec<String>) -> bool {
        let mut dp: Vec<bool> = vec![false; s.len() + 1];
        dp[s.len()] = true;
        let s = s.chars().collect::<Vec<char>>();

        for i in (0..s.len()).rev() {
            for w in word_dict.iter() {
                if i + w.len() <= s.len() && s[i..(i + w.len())].iter().collect::<String>() == *w {
                    dp[i] = dp[i + w.len()];
                }
                if dp[i] {
                    break;
                }
            }
        }

        dp[0]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_word_break_case_1() {
        // arrange
        let s = "leetcode".to_string();
        let word_dict = vec!["leet".to_string(), "code".to_string()];
        let expected = true;

        // act
        let result = Solution::word_break(s, word_dict);

        // assert
        assert_eq!(result, expected);
    }

    #[test]
    fn test_word_break_case_2() {
        // arrange
        let s = "applepenapple".to_string();
        let word_dict = vec!["apple".to_string(), "pen".to_string(), "apple".to_string()];
        let expected = true;

        // act
        let result = Solution::word_break(s, word_dict);

        // assert
        assert_eq!(result, expected);
    }

    #[test]
    fn test_word_break_case_3() {
        // arrange
        let s = "catsandog".to_string();
        let word_dict = vec![
            "cats".to_string(),
            "dog".to_string(),
            "sand".to_string(),
            "and".to_string(),
            "cat".to_string(),
        ];
        let expected = false;

        // act
        let result = Solution::word_break(s, word_dict);

        // assert
        assert_eq!(result, expected);
    }

    #[test]
    fn test_word_break_case_4() {
        // arrange
        let s = "abcd".to_string();
        let word_dict = vec![
            "a".to_string(),
            "abc".to_string(),
            "b".to_string(),
            "cd".to_string(),
        ];
        let expected = true;

        // act
        let result = Solution::word_break(s, word_dict);

        // assert
        assert_eq!(result, expected);
    }
}
